presto-spi/src/main/java/io/prestosql/spi/type/TimeZoneKey.java

```diff
@@ -210,102 +251,89 @@ public static boolean isUtcZoneId(String zoneId)
         return normalizeZoneId(zoneId).equals("utc");
     }
 
-    private static String normalizeZoneId(String originalZoneId)
+    private static String normalizeZoneId(String zoneId)
     {
-        String zoneId = originalZoneId.toLowerCase(ENGLISH);
-
-        boolean startsWithEtc = zoneId.startsWith("etc/");
-        if (startsWithEtc) {
-            zoneId = zoneId.substring(4);
-        }
-
         if (isUtcEquivalentName(zoneId)) {
             return "utc";
         }
 
-        //
-        // Normalize fixed offset time zones.
-        //
+        // The JDK doesn't understand legacy timezones such as Etc/GMT+00:00, GMT-7
+        String lowerCased = zoneId.toLowerCase(ENGLISH);
+        for (String prefix : UTC_OFFSET_PREFIXES) {
+            if (lowerCased.startsWith(prefix)) {
+                int offset = prefix.length();
+                if (lowerCased.charAt(offset) != '+' && lowerCased.charAt(offset) != '-') {
+                    // It must be something like GMT7, or Etc/UTC4, which are invalid
+                    return zoneId;
+                }
+
+                int colon = lowerCased.indexOf(':', offset);
+                int hour;
+                int minute = 0;
+                if (colon != -1) {
+                    hour = Integer.parseInt(lowerCased, offset, colon, 10);
+                    minute = Integer.parseInt(lowerCased, colon + 1, lowerCased.length(), 10);
+                }
+                else if (lowerCased.length() - offset <= 3) {
+                    // +H or +HH
+                    hour = Integer.parseInt(lowerCased, offset, lowerCased.length(), 10);
+                }
+                else {
+                    // let the JDK handle it below
+                    break;
+                }
+
+                if (hour == 0 && minute == 0) {
+                    return "utc";
+                }
+
+                if (prefix.equals("etc/gmt")) {
+                    hour = -hour;
+                }
 
-        // In some zones systems, these will start with UTC, GMT or UT.
-        int length = zoneId.length();
-        boolean startsWithEtcGmt = false;
-        if (length > 3 && (zoneId.startsWith("utc") || zoneId.startsWith("gmt"))) {
-            if (startsWithEtc && zoneId.startsWith("gmt")) {
-                startsWithEtcGmt = true;
+                return formatZoneOffset(hour, minute);
             }
-            zoneId = zoneId.substring(3);
-            length = zoneId.length();
-        }
-        else if (length > 2 && zoneId.startsWith("ut")) {
-            zoneId = zoneId.substring(2);
-            length = zoneId.length();
         }
 
-        // (+/-)00:00 is UTC
-        if ("+00:00".equals(zoneId) || "-00:00".equals(zoneId)) {
-            return "utc";
+        // Normalize using JDK rules. In particular, normalize forms such as +0735, +7, +073521
```

Comment:
Do you have a test for +073521 case? 
(what does it mean actually?)

Reply:
We don't, because we don't support them in Presto. The comment is there for completeness. +073521 is +07:35:21 (i.e., an offset with non-zero seconds)